### DISCUSSION TITLE
How to use TexText in Python scripts

### DISCUSSION THREAD

**Discussion by jcwinkler** [textext/textext#311] [Q&A] [answered]
I have an Inkscape svg file containing several TexText generated SVG snippets. I would like to change the content of these snippets from a Python script. How can this be done. (I got this question via email)

**Comment by ttrindade80** [comment]
I tested and it worked for the first time, after I rebooted my machine and ran the script. The first time I had used a svg file with only latex objects, and now, I have used one svg file with other graphical elements.
The error message:

`Traceback (most recent call last):
  File "/usr/share/inkscape/extensions/textext/base.py", line 558, in tex_to_pdf
    exec_command([tex_command, self.tmp('tex')] + self.LATEX_OPTIONS)
  File "/usr/share/inkscape/extensions/textext/utility.py", line 240, in exec_command
    raise TexTextCommandFailed(message="Command %s failed (code %d)" % (' '.join(cmd), p.returncode),
textext.errors.TexTextCommandFailed: Command /usr/bin/pdflatex tmp.tex -interaction=nonstopmode -halt-on-error failed (code 1)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/tiago/Documentos/svg_test/indutor_serie.py", line 5, in <module>
    tt.run([
  File "/usr/share/inkscape/extensions/inkex/base.py", line 131, in run
    self.save_raw(self.effect())
  File "/usr/share/inkscape/extensions/textext/base.py", line 315, in effect
    self.do_convert(new_text,
  File "/usr/share/inkscape/extensions/textext/base.py", line 388, in do_convert
    converter.tex_to_pdf(tex_executable, text, preamble_file)
  File "/usr/share/inkscape/extensions/textext/base.py", line 562, in tex_to_pdf
    raise TexTextConversionError(parsed_log, error.return_code, error.stdout, error.stderr)
textext.errors.TexTextConversionError: ! LaTeX Error: There's no line here to end.
See the LaTeX manual or LaTeX Companion for explanation.`

I tried to do this with the same latex code in the original svg file, but the error is the same.

**Comment by ttrindade80** [comment]
Really the error is in latex interpretation, if I use only a text in latex code works, but when I place math code I get the error, the latex code that I tried to use is:
`r"--text=\\begin{align*}\n\\centering\n\\SI{28}{\\henry}\n\\end{align*}",`

in the original file:
`ns3:text="\\begin{align*}\n\\centering\n\\SI{28}{\\henry}\n\\end{align*}"`

**Reply by jcwinkler** [reply]
Ok, thanks. Seems like a quotation mark problem. I will investigate this.

**Reply by ttrindade80** [reply]
The problem is with double backslash, when the TexText class tries to execute a pdflatex command in text_to _pdf does not translate the 'text' to latex format syntax. If a remove the double backslash and newline code from python string syntax.
Then, this code works:

`r"--text=\begin{align*}\n\centering\SI{100}{\henry}\end{align*}",`

But all latex code is placed only in one line, which is not very interesting to read in the TexText interface.

I'm enjoying your help, if you can help me solve this issue, I'll be very grateful.

**Reply by jcwinkler** [reply]
OK, it is my error. We do not need to pass the options as raw strings (`r` in front of the strings). I updated my original post and it is working now on my machine. Also, new lines are visible now in the TexText window. Can you confirm this @ttrindade80 ?

**Reply by ttrindade80** [reply]
Jan,

works! Thank you!

**Reply by jcwinkler** [reply]
OK, good to know @ttrindade80. I reposted the correct answer just below and it would be nice if you could mark it as answer und vote for it. Thanks!

**Comment by jcwinkler** [comment, marked as answer]
This is possible but some handwork is required. Let's assume that the file containing the TexText snippets is called `myfile.svg` and we would like to change the LaTeX-snippet
```latex
\begin{align*}
\dot x &= f(x)\\
y &= g(x)
\end{align*}
```
to
```latex
\begin{align*}
\dot x &= f(x,u)\\
y &= g(x,u)
\end{align*}
```

The following steps have to be performed:

1. Make sure `myfile.svg`is cloed in Inkscape
2. Open `myfile.svg` in an editor and locate the TexText snippets. For this purpose search for svg attributes like `texconverter` or `jacobian_root`. Here is an example:
 ```svg
       <g
       transform="matrix(0.352778,0,0,0.352778,39.975563,34.962698)"
       ns13:version="1.5.0"
       ns13:texconverter="pdflatex"
       ns13:pdfconverter="inkscape"
       ns13:text="\\begin{align*}\n\\dot x &amp;= f(x)\\\\\ny &amp;= g(x)\n\\end{align*}"
       ns13:preamble="/home/username/.config/inkscape/extensions/textext/default_packages.tex"
       ns13:scale="1.0"
       ns13:alignment="middle center"
       ns13:stroke-to-path="0"
       ns13:jacobian_sqrt="0.352778"
       id="g90">
   ```
3. For each of this snippets write down the value of the attribute `id`. Here we have `id="g90"`
4. In a Python script you can access this snippet via its id and change its content as sketched below:
    ```python
    import os
    import textext.base as textext

    tt = textext.TexText()

    tt.run([
        "--id=g90",
        "--text=\\begin{align*}\n\\dot x &= f(x,u)\\\\\ny &= g(x,u)\n\\end{align*}",
        "--preamble-file=/home/USERNAME/.config/inkscape/extensions/textext/default_packages.tex",
        "--alignment=middle center",
        "myfile.svg"], output=os.devnull)

    svg_new = "myfile_changed.svg"

    tt.document.write(svg_new)
    ```

    For this to work four things are important:
    - Replace each `\` in your original new LaTeX code by `\\`. For example `\begin{align*}`--> `\\begin{align*}` or `\\` --> `\\\\` in the `--text`option
    - A newline in the TexText editor is inserted by placing `\n` in the `--text` option
    - The path to the preamble file must be given as an absolute path
    - Inkscape's extension subsystem and the TexText extension (the directory containing the directory `textext`) is in the Python path.

5. The remaining snippets can be changed as in step 4, so a for loop makes sense which produces as new svg file from the previously created one. Note that the ids are not changed during this procedure, so you do not need to identify them again.
